### PR TITLE
[Ready] track arc_map_token during rnnt decoding

### PR DESCRIPTION
**Pull request by glynpu**
## Updated on Nov. 30, 2022:
k2.mwer is already merged by https://github.com/k2-fsa/k2/pull/1103

This pr will focus on tracking arc_map_token during decoding.

## Original:

An implementation of MWER, equation 2 of https://arxiv.org/pdf/2106.02302.pdf

Also solving issue https://github.com/k2-fsa/k2/issues/1061

## An example usage is:
1. slightly modification of [fast_beam_search](https://github.com/k2-fsa/icefall/blob/c30b8d3a1c095671d89321323390f8858722bcd1/egs/librispeech/ASR/transducer_stateless/beam_search.py#L226) to return log_probs of each time step.
```

log_probs_list = []
for t in range(T):
    *****(code of decoding)
    log_probs = (logits / temperature).log_softmax(dim=-1) 
    log_probs_list.append(log_probs)
    decoding_streams.advance(log_probs)
decoding_streams.terminate_and_flush_to_streams()
log_probs = torch.cat(log_probs_list) 
lattice, arc_map_token = decoding_streams.format_output(encoder_out_lens.tolist(), allow_partial=False)
return lattice, log_probs, arc_map_token 
```
2. Compute MWER based on the `lattice`, `log_probs`, `arc_map_token` 
```
lattice, log_probs, arc_map_token = fast_beam_search(                                                                                                                       
    model=model,                                                                                                                                                            
    decoding_graph=decoding_graph,                                                                                                                                          
    encoder_out=encoder_out,                                                                                                                                                
    encoder_out_lens=encoder_out_lens,                                                                                                                                      
    beam=beam,                                                                                                                                                              
    max_states=max_states,                                                                                                                                                  
    max_contexts=max_contexts,                                                                                                                                              
    temperature=temperature,                                                                                                                                                
)  
mwer_loss = k2.mwer_loss(nbest_scale, num_paths, log_probs, lattice, arc_map_token, ref_texts)
mwer_loss.backward()
```

## Unittest
Now I only manually check whether the gradient backprogates to the correct tokens, and it seems no obvious bugs.
Need more checks and unittests about this.
```
        # After loss.backward(),
        # users could manually check if the grad go to the right place,
        # An example:
        # By comapring hyps.labels and torch.where(log_probs.grad != 0)[1],
        # at least we could say that gradients seems to backpropate to the correct token,
        # i.e. 0, 15, 33, 10, 101 in following examples.
        # But why does token 95 in hyps.lables fail to get a gradient?
        # hyps.labels
        # >>>  tensor([  0,   0,  15,   0,  -1,   0,   0,  33,   0,  -1,   0,   0,  95,   0,
        #               -1,   0,   0, 101,   0,  -1,   0,   0,  10,   0,   0,  10,   0,  -1,
        #                0,   0,  10,   0,   0,  10,   0,   0,  10,   0,  -1],
        #             device='cuda:0', dtype=torch.int32)
        #
        # torch.where(log_probs.grad != 0)[0]  # corresopnds to context_index
        # >>>  tensor([ 0,  1,  1,  5,  6,  9, 10, 13, 14, 17, 18, 21, 22, 25, 26, 29, 30, 33,
        #               34, 38, 39, 39, 40, 44, 45, 46, 48, 49, 51, 52], device='cuda:0')
        #
        # torch.where(log_probs.grad != 0)[1]  # corresponds to token
        # >>>  tensor([  0,   0,  15,   0,   0,  33,   0,   0,   0,   0,   0,   0,   0,   0,
        #                0,   0,   0,   0,   0,  10,  10, 101,  10,  10,   0,  10,   0,   0,
        #                0,   0], device='cuda:0')
```

### CODE REVIEW

**Comment by danpovey** [comment]
In my opinion this is not really in the correct spirit of implementing backprop.  What we should be doing, I think, is replacing the scores in the lattice with scores looked-up, via arc_map_token, from the log_probs tensor, and from the decoding graph.  [I'm not sure of the details of the backprop; possibly just assigning the tensor would work... see what we do in functions for existing FSA operations]  These scores should have the same numerical values, but should have the correct backprop information.
Then all kinds of lattice/FSA operations-- not just this specific one-- would work correctly.

**Comment by glynpu** [comment]
> replacing the scores in the lattice with scores looked-up, via arc_map_token, from the log_probs tensor, and from the decoding graph

It's a good idea and helps me figure out a way to do "unittest".

## An example usage now: 
1. slightly modification of [fast_beam_search](https://github.com/k2-fsa/icefall/blob/c30b8d3a1c095671d89321323390f8858722bcd1/egs/librispeech/ASR/transducer_stateless/beam_search.py#L226) to get a lattice whose scores are  already **tracked by auto-grad machine.**

```

log_probs_list = []
t2stream_row_splits = [0]
stream2context_row_splits = [0]
num_log_probs = 0 
for t in range(T):
    *****(code of decoding)
    log_probs = (logits / temperature).log_softmax(dim=-1) 
    log_probs_list.append(log_probs)
    t2stream_row_splits += [shape.tot_size(0) + t2stream_row_splits[-1]]                                                                                                
    stream2context_row_splits += (shape.row_splits(1) + num_log_probs)[1:].tolist()                                                                                     
    num_log_probs = stream2context_row_splits[-1] 
    decoding_streams.advance(log_probs)
decoding_streams.terminate_and_flush_to_streams()
log_probs = torch.cat(log_probs_list) 
# NOTE: now lattice.scores is already tracked by auto-grad machine.
lattice = decoding_streams.format_output(encoder_out_lens.tolist(), log_probs=log_probs, t2s2c_shape=t2stream2context_shape3)
return lattice
```

2.  `lattice` and `ref_texts` are enough to compute MWER.
     i.e. current k2.mwer loss doesn't care how the lattice generated.
     It could be generated from a rnnt model by `fast_beam_search`, 
     or from a ctc model by `intersect_dense_pruned`.   
```
mwer_loss = k2.mwer_loss(nbest_scale, num_paths, lattice, ref_texts)
mwer_loss.backward()
```

## Unittest
Now a simple test is added for scores looking up by `arc_map_token` from log_probs.
With line-277, at least we could say `arc_map_token` is generated correctly.

https://github.com/k2-fsa/k2/blob/e5a8e4e2546e65c73c3a72353c044a52fb4ef707/k2/python/k2/rnnt_decode.py#L273-L279

File: k2/python/k2/mwer_loss.py [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+
+
+        # Group each log_prob into [path][arc]
+        ragged_nbest_logp = k2.RaggedTensor(nbest.kept_path.shape, nbest_logp)
+        # Get the probalitity of each path, in log format, with shape [path].
+        path_logp = ragged_nbest_logp.sum()
+
+        # Group path_logp into [stream][path] to compute denominator of each stream.
+        # A stream here means an input wav.
+        ragged_path_prob = k2.RaggedTensor(nbest.shape, path_logp.exp())
```
Comment (danpovey) [inline k2/python/k2/mwer_loss.py:397]: You could perhaps write a softmax function accepting RaggedTensor to replace these few lines, and  call that instead, it will be clearer.

File: k2/python/k2/mwer_loss.py [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+    assert aux_labels.num_axes == 2
+    if return_ragged:
+        return aux_labels
+    else:
+        return aux_labels.tolist()
+
+
+# Copied from icefall.
+# TODO(liyong) simplify this.
+class Nbest(object):
```
Comment (danpovey) [inline k2/python/k2/mwer_loss.py:53]: If this is going to be used it should probably be in a different file.   But are you sure it's not already in k2?

File: k2/python/k2/mwer_loss.py [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+        return k2.RaggedTensor(self.shape, tot_scores)
+
+    def build_levenshtein_graphs(self) -> k2.Fsa:
+        """Return an FsaVec with axes [utt][state][arc]."""
+        word_ids = get_texts(self.fsa, return_ragged=True)
+        return k2.levenshtein_graph(word_ids)
+
+
+class MWERLoss(nn.Module):
+    # Minimus Word Error Rate.
```
Comment (danpovey) [inline k2/python/k2/mwer_loss.py:354]: I am concerned that, especially for long utterances, the dynamic range of scores will be too large and the posteriors will be mostly 0 or 1.  To prevent this it might be a good idea to have an extra argument that functions like a temperature-- something we can scale the logprobs by before doing the softmax.

**Review by danpovey — COMMENTED** [review]
some comments..

File: k2/python/k2/mwer_loss.py (hunk 1 above)
Comment (glynpu) [reply k2/python/k2/mwer_loss.py]: A function `k2.ragged_softmax` is added to k2/ops.py.

File: k2/python/k2/mwer_loss.py (hunk 2 above)
Comment (glynpu) [reply k2/python/k2/mwer_loss.py]: Yes, class `Nbest` is also exists in k2. 
But it has fewer functions than that in icefall, i.e. `from_lattice` and `build_levenshtein_graphs`. So I just copied the icefall one here. 
Now in the latest pr, this two needed functions  are added into the k2.Nbest.

**Comment by glynpu** [comment]
The most important unit test for this pr is following assertion:

https://github.com/k2-fsa/k2/blob/d83d79ae1d43b2c6e9fd486953cb5c4eb62cf1ed/k2/python/k2/rnnt_decode.py#L284-L286

`fsa.scores` is generated during decoding.
`scores_tracked_by_autograd` is selected by `arc_map_token`

If `fsa.scores` is identical to `scores_tracked_by_autograd`,  it means `arc_map_token` is tracked correctly.

File: k2/csrc/rnnt_decode.cu [hunk 4]
```diff
@@ -885,12 +952,36 @@ void RnntDecodingStreams::FormatOutput(const std::vector<int32_t> &num_frames,
...
+                    t2s2c_shape_row_splits2[t2s2c_stream_idx01];
+          // oarc_idx2 is for context.
+          int32_t oarc_idx2 = oarc_idx012 - oarc_idx01x;
+          int32_t t2s2c_context_idx012 = t2s2c_context_idx01x + oarc_idx2;
+          // Manually map arc to super_final state(i.e. arc.label == -1)
+          // to blank_id == 0.
+          int32_t arc_label = 0;
+          if (arc.label != -1) {
+            arc_label = arc.label;
+          }
```
Comment (pkufool) [inline k2/csrc/rnnt_decode.cu:972]: I think you can set the `arc_map_token` to `-1` for final arcs,  and `k2.index_select` can map `-1` to default value  automatically.

File: k2/python/k2/rnnt_decode.py [hunk 5]
```diff
@@ -173,16 +175,27 @@ def format_output(
             If false, we only care about the real final state in the
             decoding graph on the last frame when generating lattice.
             Default False.
```
Comment (pkufool) [inline k2/python/k2/rnnt_decode.py:177]: Please add documents for `log_probs`.

File: k2/python/k2/rnnt_decode.py [hunk 6]
```diff
@@ -264,4 +277,14 @@ def format_output(
             for name, value in src.named_non_tensor_attr():
                 setattr(fsa, name, value)
 
+        if log_probs is not None:
+            # Make fsa.scores tracked by autograd
+            # to make the whole decoding process differentiable.
+            scores_tracked_by_autograd = torch.index_select(
+                log_probs.reshape(-1), 0, arc_map_token)
+            final_arc_index = torch.where(fsa.arcs.values()[:, 2] == -1)
+            scores_tracked_by_autograd[final_arc_index] *= 0
```
Comment (pkufool) [inline k2/python/k2/rnnt_decode.py:286]: See the comment above, if you add `-1` to `arc_map_token` and use `k2.index_select` you can avoid line 285-286.

File: k2/csrc/rnnt_decode.cu [hunk 7]
```diff
@@ -823,55 +850,95 @@ void RnntDecodingStreams::FormatOutput(const std::vector<int32_t> &num_frames,
...
+                *t2s2c_shape_row_splits1 = nullptr;
+
+  // In `GatherPrevFrames`, we use
+  // Unstack(frames, 1, false /*pad_right*/, &prev_frames);
+  // to generated prev_frames.
+  // Since pad_right=false, shorter utterances are padded to the left side,
+  // i.e. padded beginning frames.
+  // We need to subtract number of padded frames when calculating
+  // the real time index.
+  auto num_padded_frames = Array1<int32_t>(c_, num_frames);
```
Comment (pkufool) [inline k2/csrc/rnnt_decode.cu:875]: You initialize `num_padded_frames` with real frames here, it takes me a while to get your idea till I read the following lines. Please add comments here.

**Review by pkufool — COMMENTED** [review]
LGTM

Leave some comments here. It would be great if you can run the `fast_beam_search` and `fast_beam_search_LG` in icefall with this change to double check its correctness.

File: k2/python/k2/rnnt_decode.py [hunk 8]
```diff
@@ -264,4 +277,14 @@ def format_output(
...
+            # This assertion statement is kind of unit test.
+            assert torch.all(fsa.scores == scores_tracked_by_autograd)
```
Comment (pkufool) [inline k2/python/k2/rnnt_decode.py:288]: I think it is not right if we are using an LG graph.

**Review by pkufool — APPROVED** [review]
+2

**Comment by desh2608** [comment]
> t2stream2context_shape3

@glynpu Could you tell me how you obtain `t2stream2context_shape3`?

**Comment by desh2608** [comment]
> > t2stream2context_shape3
> 
> @glynpu Could you tell me how you obtain `t2stream2context_shape3`?

Found the answer here: https://github.com/k2-fsa/k2/blob/42e92fdd4097adcfe9937b4d2df7736d227b8e85/k2/python/tests/rnnt_decode_test.py#L138
